planet/api/http.py

```diff
@@ -13,135 +13,176 @@
-                key = re.match(r'api-key (\S+)', existing_auth)
-                if key:
-                    prepared_request.prepare_url(
-                        prepared_request.url, {
-                            'api_key': key.group(1)
-                        }
-                    )
+        http_resp = await self._client.send(request.http_request,
+                                            stream=stream)
+        return models.Response(request, http_resp)
+
+    def stream(self, request):
+        '''Submit a request and get the response as a stream context manager.
+
+        :param request: Request to submit
+        :type request: planet.api.models.Request
+        :returns: Context manager providing the body as a stream.
+        :rtype: APlanetStream
+        '''
+        return APlanetStream(
+            session=self,
+            request=request
+        )
 
     @staticmethod
-    def _is_subdomain_of_tld(url1, url2):
-        orig_host = urlparse(url1).hostname
-        re_host = urlparse(url2).hostname
-        return orig_host.split('.')[-2:] == re_host.split('.')[-2:]
+    def _get_user_agent():
+        return 'planet-client-python/' + __version__
+
+    @staticmethod
+    async def _log_request(request):
+        LOGGER.info(f'{request.method} {request.url} - Sent')
+
+    @staticmethod
+    async def _log_response(response):
+        request = response.request
+        LOGGER.info(
+            f'{request.method} {request.url} - '
+            f'Status {response.status_code}')
+
+    @staticmethod
+    async def _raise_for_status(response):
+        # TODO: consider using http_response.reason_phrase
+        status = response.status_code
+
+        if status < 300:
+            return
+
+        exception = {
+            400: exceptions.BadQuery,
+            401: exceptions.InvalidAPIKey,
+            403: exceptions.NoPermission,
+            404: exceptions.MissingResource,
+            429: exceptions.TooManyRequests,
+            500: exceptions.ServerError
+        }.get(status, None)
+
+        try:
+            msg = response.text
+        except httpx.ResponseNotRead:
+            await response.aread()
+            msg = response.text
+
+        # differentiate between over quota and rate-limiting
+        if status == 429 and 'quota' in msg.lower():
+            exception = exceptions.OverQuota
+
+        if exception:
+            raise exception(msg)
+
+        raise exceptions.APIException(f'{status}: {msg}')
+
+
+class APlanetStream():
+    '''Context manager for asynchronous response stream from Planet server.
+
+    :param session: Open session to Planet server
+    :type session: APlanetSession
+    :param request: Request to submit
+    :type request: planet.api.models.Request
+    '''
+    def __init__(self, session, request):
+        self.session = session
+        self.request = request
+
+    async def __aenter__(self):
+        self.response = await self.session.request(
```

Comment:
I feel like I'm misunderstanding something here, but can all instances of this class be replaced with this request instead? It looks like it could save a bit of overhead if that's the case.

Reply:
This class manages an async response, both initiating it as a stream and closing it when the context is left. One job, and an important job.

Reply:
I don't feel like my previous question was answered. What value does this provide over directly calling `session.request(request, stream=True)`?

Reply:
calling `request.aclose()` when the context is left

Reply:
I'm not understanding your most recent response. It looks like if I were to call the request directly instead of using this class, that it would obviate the need for this class, i.e. the response is closed when you call it directly.

Reply:
@ericrdunham I don't follow your question here, circumventing the async context manager would negate the utility of async execution. This is good as-is.

For reference: https://www.python.org/dev/peps/pep-0492/#asynchronous-context-managers-and-async-with